parsers/CA_NS.py

```diff
@@ -1,213 +1,210 @@
-        # Use 98% instead.
-        "coal": (0, 0.98),
-        "gas": (0, 0.5),
-        "biomass": (0, 0.15),
-        "hydro": (0, 0.60),
-        "wind": (0, 0.55),
-        "imports": (0, 0.50),
-    }
-
-    # Sanity checks: verify that reported production doesn't exceed listed capacity by a lot.
-    # In particular, we've seen error cases where hydro production ends up calculated as 900 MW
-    # which greatly exceeds known capacity of 418 MW.
-    valid_absolute = {
-        "coal": 1300,
-        "gas": 700,
-        "biomass": 100,
-        "hydro": 500,
-        "wind": 700,
-    }
-
-    mix_url = "https://www.nspower.ca/library/CurrentLoad/CurrentMix.json"
-    mix_data = requests_obj.get(mix_url).json()
-
-    load_url = "https://www.nspower.ca/library/CurrentLoad/CurrentLoad.json"
-    load_data = requests_obj.get(load_url).json()
-
-    production = []
-    imports = []
-    for mix in mix_data:
-        percent_mix = {
-            "coal": mix["Solid Fuel"] / 100.0,
-            "gas": (mix["HFO/Natural Gas"] + mix["CT's"] + mix["LM 6000's"]) / 100.0,
-            "biomass": mix["Biomass"] / 100.0,
-            "hydro": mix["Hydro"] / 100.0,
-            "wind": mix["Wind"] / 100.0,
-            "imports": mix["Imports"] / 100.0,
+from electricitymap.contrib.config import ZoneKey
+from electricitymap.contrib.lib.models.event_lists import (
+    ExchangeList,
+    ProductionBreakdownList,
+)
+from electricitymap.contrib.lib.models.events import ProductionMix
+from parsers.lib.exceptions import ParserException
+
+LOAD_URL = "https://www.nspower.ca/library/CurrentLoad/CurrentLoad.json"
+MIX_URL = "https://www.nspower.ca/library/CurrentLoad/CurrentMix.json"
+PARSER = "CA_NS.py"
+SOURCE = "nspower.ca"
+# Sanity checks: verify that reported production doesn't exceed listed capacity
+# by a lot. In particular, we've seen error cases where hydro production ends
+# up calculated as 900 MW which greatly exceeds known capacity of 418 MW.
+MEGAWATT_LIMITS = {
+    "coal": 1300,
+    "gas": 700,
+    "biomass": 100,
+    "hydro": 500,
+    "wind": 700,
+}
+# This is based on validation logic in
+# https://www.nspower.ca/site/renewables/assets/js/site.js. In practical terms,
+# I've seen hydro production go way too high (>70%) which is way more than
+# reported capacity.
+FRACTION_LIMITS = {
+    # The validation JS reports an error when Solid Fuel (coal) is over 85%,
+    # but as far as I can tell, that can actually be a valid result, I've seen
+    # it a few times. Use 98% instead.
+    "coal": (0, 0.98),
+    "gas": (0, 0.5),
+    "biomass": (0, 0.15),
+    "hydro": (0, 0.60),
+    "wind": (0, 0.55),
+    "imports": (0, 0.50),
+}
+ZONE_KEY = ZoneKey("CA-NS")
+
+
+def _get_ns_info(
+    session: Session, logger: Logger
+) -> (ExchangeList, ProductionBreakdownList):
+    base_loads = session.get(LOAD_URL).json()  # Base loads in MW
+    mixes_percent = session.get(MIX_URL).json()  # Electricity breakdowns in %
```

Comment:
I've observed some interesting characteristics in the pair of JSON arrays returned by these endpoints:

1. The arrays $a$ and $b$ appear to be parallel in the sense that element $a_i$ corresponds with element $b_i$ for all $i$. I've convinced myself that this is true during testing by inspecting the timestamps. We can use this fact to our advantage by `zip`ping the two arrays together and iterating over the resulting `tuple`s instead of the current approach of iterating over one and then performing a linear search to find the corresponding entry in the other. Unfortunately, a race condition is lurking here: we could get unlucky and make one request immediately before a new datapoint comes in and the other immediately afterward, resulting in desynchronised arrays. We could address this problem by verifying that all timestamps agree between the two arrays; if they don't, I imagine we could simply raise an exception and bail out, and we'll get that datapoint on the next go round. Alternatively (additionally?), we could retry the requests a few times (likely after sleeping for a second or so) until the returned arrays do agree.
3. Each array is of length 25, not 24. The first (oldest, according to the timestamp) element of the base load array always indicates $0 \ \mathrm{MW}$, so I think we should just ignore this entry in both arrays.

Reply:
I went ahead and implemented this in a second commit!